src/log.c

```diff
@@ -55,54 +68,49 @@ void log_message_internal(const char* tag, log_level_t level, const char* messag
     fprintf(out, "\n");
     fflush(out);
 }
-void log_message_vargs(log_producer_t *tag, log_level_t level, const char *message, va_list args) {
-    for(int i =0; i<= log_service->size-1; ++i){
-        log_sink_t* sink = log_service->sinks[i];
-        if (level > sink->min_level) {
+
+void log_message_internal(log_producer_t *producer, log_level_t level, const char *message, va_list args) {
+    for(uint8_t i = 0; i < log_sinks_registered_count; ++i){
+        log_sink_t* sink = &log_sinks_registered_list[i];
+        if ((level & sink->levels) != level) {
             continue;
         }
-        log_message_internal(tag->tag, level, message, args,sink->out);
+
+        log_message_internal_printer(producer->tag, level, message, args,sink->out);
     }
 }
-void log_message(log_producer_t* tag, log_level_t level, const char* message, ...) {
+
+void log_message(log_producer_t* producer, log_level_t level, const char* message, ...) {
     va_list args;
     va_start(args, message);
 
-    log_message_vargs(tag,level,message,args);
+    log_message_internal(producer, level,message,args);
 
     va_end(args);
 }
 
-void __attribute__((destructor)) deinit_log_service() {
-    free(log_service->sinks);
-    free(log_service);
-}
-
-void __attribute__((constructor)) init_log_service() {
-    log_service = (log_service_t*)malloc(sizeof(log_service));
-    //init console out
-    log_sink_t* console = init_log_sink(stdout, LOG_LEVEL_INFO);
-    log_service->sinks = malloc(sizeof(log_sink_t));
-    *log_service->sinks = console;
-    log_service->size = 1;
-}
-
 void log_sink_register(log_sink_t *sink) {
-    ++log_service->size;
-    log_service->sinks = realloc(log_service->sinks, sizeof(log_sink_t)*log_service->size);
-    *(log_service->sinks + log_service->size - 1) = sink;
+    log_sinks_registered_list[log_sinks_registered_count] = *sink;
```

Comment:
It should be `log_sinks_registered_list[log_sinks_registered_count - 1]` otherewise it will raise error.